File: backend/apps/network/calculations.py
from __future__ import annotations

from dataclasses import dataclass, field
from math import ceil


@dataclass
class NetworkError:
    work_id: int | None
    field: str
    message: str


@dataclass
class NetworkWork:
    id: int
    work_number: str
    event_start: str
    event_end: str
    title: str
    labor: int
    workers: int
    duration: int
    calculation_mode: str = "fixed_labor"
    planned_shift: int = 0
    order_index: int = 0
    dependencies: list[int] = field(default_factory=list)
    early_start: int = 0
    early_finish: int = 0
    late_start: int = 0
    late_finish: int = 0
    total_float: int = 0
    is_critical: bool = False


def normalize_work(work: NetworkWork) -> NetworkWork:
    if work.calculation_mode == "fixed_labor":
        work.duration = max(1, ceil(work.labor / work.workers))
    else:
        work.duration = max(1, work.duration)
        work.labor = max(1, work.duration * work.workers)
    return work


def build_dependencies_from_events(works: list[NetworkWork]) -> list[NetworkWork]:
    for work in works:
        work.dependencies = [
            other.id
            for other in works
            if other.id != work.id and other.event_end == work.event_start
        ]
    return works


def validate_network(works: list[NetworkWork]) -> list[NetworkError]:
    errors: list[NetworkError] = []
    if not works:
        return [NetworkError(None, "project", "Проект должен иметь хотя бы одну работу.")]
    for work in works:
        if not work.work_number:
            errors.append(NetworkError(work.id, "work_number", "Номер работы не должен быть пустым."))
        if not work.title:
            errors.append(NetworkError(work.id, "title", "Заполните название работы."))
        if not work.event_start:
            errors.append(NetworkError(work.id, "event_start", "Заполните событие начала."))
        if not work.event_end:
            errors.append(NetworkError(work.id, "event_end", "Заполните событие окончания."))
        if work.event_start and work.event_start == work.event_end:
            errors.append(NetworkError(work.id, "event_end", "Событие начала и событие окончания не могут совпадать."))
        if work.labor < 1:
            errors.append(NetworkError(work.id, "labor", "Трудоёмкость должна быть больше 0."))
        if work.workers < 1:
            errors.append(NetworkError(work.id, "workers", "Количество исполнителей должно быть больше 0."))
        if work.duration < 1:
            errors.append(NetworkError(work.id, "duration", "Продолжительность должна быть больше 0."))
    if not errors and detect_cycles(works):
        errors.append(NetworkError(None, "project", "В графике обнаружен цикл. Проверьте события начала и окончания."))
    return errors


def detect_cycles(works: list[NetworkWork]) -> bool:
    return len(_topological_sort(works)) != len(works)


def calculate_schedule(works: list[NetworkWork]) -> tuple[list[NetworkWork], list[NetworkError]]:
    works = build_dependencies_from_events([normalize_work(work) for work in works])
    errors = validate_network(works)
    if errors:
        return works, errors
    by_id = {work.id: work for work in works}
    successors = _successors(works)
    order = _topological_sort(works)
    base_by_id = {work.id: _clone_work(work) for work in works}
    base = _forward_schedule(order, base_by_id, 0)
    base_duration = max((work.early_finish for work in base), default=0)
    _apply_late_dates(list(reversed(order)), base_by_id, successors, base_duration)
    if not any(work.planned_shift > 0 for work in works):
        return [base_by_id[work_id] for work_id in order], []

    for work_id in order:
        work = by_id[work_id]
        dependency_finish = max([by_id[dependency].early_finish for dependency in work.dependencies] or [0])
        allowed_shift = max(0, base_by_id[work_id].late_start - dependency_finish)
        work.planned_shift = min(max(0, work.planned_shift), allowed_shift)
        work.early_start = dependency_finish + work.planned_shift
        work.early_finish = work.early_start + work.duration
    project_duration = max((work.early_finish for work in works), default=0)
    _apply_late_dates(list(reversed(order)), by_id, successors, project_duration)
    return [by_id[work_id] for work_id in order], []


def _clone_work(work: NetworkWork) -> NetworkWork:
    return NetworkWork(
        work.id,
        work.work_number,
        work.event_start,
        work.event_end,
        work.title,
        work.labor,
        work.workers,
        work.duration,
        work.calculation_mode,
        0,
        work.order_index,
        list(work.dependencies),
    )


def _forward_schedule(order: list[int], by_id: dict[int, NetworkWork], default_shift: int) -> list[NetworkWork]:
    for work_id in order:
        work = by_id[work_id]
        work.planned_shift = default_shift
        work.early_start = max([by_id[dependency].early_finish for dependency in work.dependencies] or [0]) + default_shift
        work.early_finish = work.early_start + work.duration
    return [by_id[work_id] for work_id in order]


def _apply_late_dates(order: list[int], by_id: dict[int, NetworkWork], successors: dict[int, list[int]], project_duration: int):
    for work_id in order:
        work = by_id[work_id]
        next_ids = successors.get(work_id, [])
        work.late_finish = min([by_id[next_id].late_start for next_id in next_ids] or [project_duration])
        work.late_start = work.late_finish - work.duration
        work.total_float = work.late_start - work.early_start
        work.is_critical = work.total_float == 0


def calculate_resource_usage(works: list[NetworkWork], resource_limit: int) -> list[dict]:
    duration = max((work.early_finish for work in works), default=0)
    usage = []
    for day in range(duration):
        workers = sum(work.workers for work in works if day >= work.early_start and day < work.early_finish)
        usage.append({
            "day": day + 1,
            "workers": workers,
            "overloaded": workers > resource_limit,
            "idle": workers > 0 and workers < resource_limit * 0.45,
        })
    return usage


def calculate_resource_usage_with_float(works: list[NetworkWork], resource_limit: int) -> list[dict]:
    duration = max((work.early_finish for work in works), default=0)
    allocations = [_reserve_aware_allocation(work) for work in works]
    usage = []
    for day in range(duration):
        workers = sum(allocation["workers"] for allocation in allocations if day >= allocation["start"] and day < allocation["finish"])
        usage.append({
            "day": day + 1,
            "workers": workers,
            "overloaded": workers > resource_limit,
            "idle": workers > 0 and workers < resource_limit * 0.45,
        })
    return usage


def _reserve_aware_allocation(work: NetworkWork) -> dict:
    available_duration = max(1, work.duration + max(0, work.total_float))
    workers = max(1, ceil(work.labor / available_duration)) if work.total_float > 0 else work.workers
    duration = min(available_duration, max(1, ceil(work.labor / workers))) if work.total_float > 0 else work.duration
    return {
        "start": work.early_start,
        "finish": min(work.late_finish, work.early_start + duration),
        "workers": workers,
    }


def calculate_metrics(works: list[NetworkWork], resource_limit: int) -> dict:
    usage = calculate_resource_usage(works, resource_limit)
    project_duration = max((work.early_finish for work in works), default=0)
    max_workers = max((point["workers"] for point in usage), default=0)
    average_workers = sum(point["workers"] for point in usage) / len(usage) if usage else 0
    return {
        "project_duration": project_duration,
        "max_workers": max_workers,
        "average_workers": round(average_workers, 2),
        "critical_works_count": sum(1 for work in works if work.is_critical),
        "works_with_float_count": sum(1 for work in works if work.total_float > 0),
        "overload_count": sum(1 for point in usage if point["overloaded"]),
    }


def _successors(works: list[NetworkWork]) -> dict[int, list[int]]:
    result = {work.id: [] for work in works}
    for work in works:
        for dependency_id in work.dependencies:
            result.setdefault(dependency_id, []).append(work.id)
    return result


def _topological_sort(works: list[NetworkWork]) -> list[int]:
    incoming = {work.id: 0 for work in works}
    successors = _successors(works)
    for work in works:
        incoming[work.id] = len(work.dependencies)
    queue = [work.id for work in works if incoming[work.id] == 0]
    order: list[int] = []
    while queue:
        work_id = queue.pop(0)
        order.append(work_id)
        for next_id in successors.get(work_id, []):
            incoming[next_id] -= 1
            if incoming[next_id] == 0:
                queue.append(next_id)
    return order

File: src/domain/network/calculations.ts
import type {
  NetworkValidationError,
  ProjectMetrics,
  ResourceUsagePoint,
  ScheduleResult,
  ScheduledWorkItem,
  WorkDefinition,
  WorkItem,
} from './types';

const idleThreshold = 0.45;

export function recalculateDurationsByLabor(workItems: WorkItem[]): WorkItem[] {
  return workItems.map((item) => ({
    ...item,
    duration: Math.max(1, Math.ceil(item.labor / item.workers)),
  }));
}

export function recalculateDefinition(definition: WorkDefinition): WorkDefinition {
  const labor = sanitizePositiveNumber(definition.labor);
  const workers = sanitizePositiveNumber(definition.workers);
  const duration =
    definition.calculationMode === 'manual-duration'
      ? sanitizePositiveNumber(definition.duration)
      : Math.max(1, Math.ceil(labor / workers));

  return {
    ...definition,
    labor,
    workers,
    duration,
  };
}

export function buildWorkItemsFromDefinitions(definitions: WorkDefinition[]): { items: WorkItem[]; errors: NetworkValidationError[] } {
  const errors = validateWorkDefinitions(definitions);
  if (errors.length > 0) {
    return { items: [], errors };
  }

  const normalized = definitions.map(recalculateDefinition);
  const items = normalized.map((definition) => ({
    id: definition.id,
    code: definition.code.trim(),
    title: definition.title.trim(),
    from: definition.from.trim(),
    to: definition.to.trim(),
    labor: definition.labor,
    workers: definition.workers,
    duration: definition.duration,
    plannedShift: definition.plannedShift,
    dependencies: normalized
      .filter((candidate) => candidate.id !== definition.id && candidate.to.trim() === definition.from.trim())
      .map((candidate) => candidate.id),
  }));

  return { items, errors: validateNetwork(items) };
}

export function validateWorkDefinitions(definitions: WorkDefinition[]): NetworkValidationError[] {
  const errors: NetworkValidationError[] = [];

  if (definitions.length === 0) {
    errors.push({ type: 'empty-project', field: 'project', message: 'Проект должен иметь хотя бы одну работу.' });
    return errors;
  }

  for (const definition of definitions) {
    if (!definition.code.trim()) {
      errors.push({ type: 'empty-code', field: 'code', workId: definition.id, message: 'Номер работы не должен быть пустым.' });
    }
    if (!definition.title.trim()) {
      errors.push({ type: 'empty-title', field: 'title', workId: definition.id, message: 'Заполните название работы.' });
    }
    if (!definition.from.trim()) {
      errors.push({ type: 'empty-from', field: 'from', workId: definition.id, message: 'Заполните событие начала.' });
    }
    if (!definition.to.trim()) {
      errors.push({ type: 'empty-to', field: 'to', workId: definition.id, message: 'Заполните событие окончания.' });
    }
    if (definition.from.trim() && definition.from.trim() === definition.to.trim()) {
      errors.push({
        type: 'same-event',
        field: 'to',
        workId: definition.id,
        message: 'Событие начала и событие окончания не могут совпадать.',
      });
    }
    if (!Number.isFinite(definition.labor) || definition.labor < 1) {
      errors.push({ type: 'invalid-labor', field: 'labor', workId: definition.id, message: 'Трудоёмкость должна быть больше 0.' });
    }
    if (!Number.isFinite(definition.workers) || definition.workers < 1) {
      errors.push({
        type: 'invalid-workers',
        field: 'workers',
        workId: definition.id,
        message: 'Количество исполнителей должно быть больше 0.',
      });
    }
    if (!Number.isFinite(definition.duration) || definition.duration < 1) {
      errors.push({
        type: 'invalid-duration',
        field: 'duration',
        workId: definition.id,
        message: 'Продолжительность должна быть больше 0.',
      });
    }
  }

  return errors;
}

export function updateWorkWorkers(workItems: WorkItem[], workItemId: string, workers: number): WorkItem[] {
  if (workers < 1) {
    throw new Error('Количество сотрудников должно быть не меньше 1.');
  }

  return workItems.map((item) =>
    item.id === workItemId
      ? {
          ...item,
          workers,
          duration: Math.max(1, Math.ceil(item.labor / workers)),
        }
      : item,
  );
}

export function updateWorkDuration(workItems: WorkItem[], workItemId: string, duration: number): WorkItem[] {
  if (duration < 1) {
    throw new Error('Длительность должна быть не меньше 1 дня.');
  }

  return workItems.map((item) =>
    item.id === workItemId
      ? {
          ...item,
          duration,
          workers: Math.max(1, Math.ceil(item.labor / duration)),
        }
      : item,
  );
}

export function shiftWork(workItems: WorkItem[], workItemId: string, shift: number): WorkItem[] {
  return workItems.map((item) => (item.id === workItemId ? { ...item, plannedShift: Math.max(0, shift) } : item));
}

export function calculateSchedule(workItems: WorkItem[]): ScheduleResult {
  const errors = validateNetwork(workItems);
  if (errors.length > 0) {
    return { items: [], projectDuration: 0, errors };
  }

  const order = topologicalSort(workItems);
  const byId = new Map(workItems.map((item) => [item.id, item]));
  const successors = buildSuccessors(workItems);
  const baseSchedule = buildForwardSchedule(order, byId, 0);
  const baseProjectDuration = getScheduleDuration(baseSchedule);
  applyLateDates(order, baseSchedule, successors, baseProjectDuration);
  const hasManualShifts = order.some((id) => (byId.get(id)?.plannedShift ?? 0) > 0);

  if (!hasManualShifts) {
    return { items: order.map((id) => baseSchedule.get(id)!), projectDuration: baseProjectDuration, errors: [] };
  }

  const shiftedSchedule = new Map<string, ScheduledWorkItem>();
  for (const id of order) {
    const item = byId.get(id)!;
    const baseItem = baseSchedule.get(id)!;
    const dependencyFinish = item.dependencies.reduce((finish, dependencyId) => {
      return Math.max(finish, shiftedSchedule.get(dependencyId)?.earlyFinish ?? 0);
    }, 0);
    const requestedShift = Math.max(0, item.plannedShift ?? 0);
    const allowedShift = Math.max(0, baseItem.lateStart - dependencyFinish);
    const appliedShift = Math.min(requestedShift, allowedShift);
    const earlyStart = dependencyFinish + appliedShift;
    const earlyFinish = earlyStart + Math.max(1, item.duration);

    shiftedSchedule.set(id, {
      ...item,
      plannedShift: appliedShift,
      duration: Math.max(1, item.duration),
      earlyStart,
      earlyFinish,
      lateStart: 0,
      lateFinish: 0,
      totalFloat: 0,
      freeFloat: 0,
      isCritical: false,
    });
  }

  const shiftedProjectDuration = getScheduleDuration(shiftedSchedule);
  applyLateDates(order, shiftedSchedule, successors, shiftedProjectDuration);

  return { items: order.map((id) => shiftedSchedule.get(id)!), projectDuration: shiftedProjectDuration, errors: [] };
}

function buildForwardSchedule(order: string[], byId: Map<string, WorkItem>, defaultShift: number): Map<string, ScheduledWorkItem> {
  const scheduled = new Map<string, ScheduledWorkItem>();
  for (const id of order) {
    const item = byId.get(id)!;
    const dependencyFinish = item.dependencies.reduce((finish, dependencyId) => {
      return Math.max(finish, scheduled.get(dependencyId)?.earlyFinish ?? 0);
    }, 0);
    const earlyStart = dependencyFinish + defaultShift;
    const earlyFinish = earlyStart + Math.max(1, item.duration);

    scheduled.set(id, {
      ...item,
      plannedShift: defaultShift,
      duration: Math.max(1, item.duration),
      earlyStart,
      earlyFinish,
      lateStart: 0,
      lateFinish: 0,
      totalFloat: 0,
      freeFloat: 0,
      isCritical: false,
    });
  }
  return scheduled;
}

function getScheduleDuration(scheduled: Map<string, ScheduledWorkItem>): number {
  return scheduled.size ? Math.max(...Array.from(scheduled.values()).map((item) => item.earlyFinish)) : 0;
}

function applyLateDates(
  order: string[],
  scheduled: Map<string, ScheduledWorkItem>,
  successors: Map<string, string[]>,
  projectDuration: number,
) {
  for (const id of [...order].reverse()) {
    const item = scheduled.get(id)!;
    const itemSuccessors = successors.get(id) ?? [];
    const lateFinish =
      itemSuccessors.length === 0
        ? projectDuration
        : Math.min(...itemSuccessors.map((successorId) => scheduled.get(successorId)!.lateStart));
    item.lateFinish = lateFinish;
    item.lateStart = lateFinish - item.duration;
    item.totalFloat = item.lateStart - item.earlyStart;
    item.freeFloat =
      itemSuccessors.length === 0
        ? projectDuration - item.earlyFinish
        : Math.min(...itemSuccessors.map((successorId) => scheduled.get(successorId)!.earlyStart)) - item.earlyFinish;
    item.isCritical = item.totalFloat === 0;
  }
}

export function calculateResourceUsage(workItems: WorkItem[], resourceLimit = 0): ResourceUsagePoint[] {
  const schedule = isScheduled(workItems) ? { items: workItems as ScheduledWorkItem[] } : calculateSchedule(workItems);
  if (!schedule.items.length) {
    return [];
  }

  const duration = Math.max(...schedule.items.map((item) => item.earlyFinish));
  return Array.from({ length: duration }, (_, day) => {
    const workers = schedule.items.reduce((sum, item) => {
      return day >= item.earlyStart && day < item.earlyFinish ? sum + item.workers : sum;
    }, 0);

    return {
      day: day + 1,
      workers,
      overloaded: resourceLimit > 0 && workers > resourceLimit,
      idle: resourceLimit > 0 && workers > 0 && workers < resourceLimit * idleThreshold,
    };
  });
}

export function calculateResourceUsageWithFloat(workItems: WorkItem[], resourceLimit = 0): ResourceUsagePoint[] {
  const schedule = isScheduled(workItems) ? { items: workItems as ScheduledWorkItem[] } : calculateSchedule(workItems);
  if (!schedule.items.length) {
    return [];
  }

  const duration = Math.max(...schedule.items.map((item) => item.earlyFinish));
  const allocations = schedule.items.map(getReserveAwareAllocation);

  return Array.from({ length: duration }, (_, day) => {
    const workers = allocations.reduce((sum, allocation) => {
      return day >= allocation.start && day < allocation.finish ? sum + allocation.workers : sum;
    }, 0);

    return {
      day: day + 1,
      workers,
      overloaded: resourceLimit > 0 && workers > resourceLimit,
      idle: resourceLimit > 0 && workers > 0 && workers < resourceLimit * idleThreshold,
    };
  });
}

function getReserveAwareAllocation(item: ScheduledWorkItem) {
  const availableDuration = Math.max(1, item.duration + Math.max(0, item.totalFloat));
  const workers = item.totalFloat > 0 ? Math.max(1, Math.ceil(item.labor / availableDuration)) : item.workers;
  const duration = item.totalFloat > 0 ? Math.min(availableDuration, Math.max(1, Math.ceil(item.labor / workers))) : item.duration;

  return {
    start: item.earlyStart,
    finish: Math.min(item.lateFinish, item.earlyStart + duration),
    workers,
  };
}

export function getCriticalPath(workItems: WorkItem[]): ScheduledWorkItem[] {
  const schedule = isScheduled(workItems) ? (workItems as ScheduledWorkItem[]) : calculateSchedule(workItems).items;
  return schedule.filter((item) => item.isCritical);
}

export function getProjectMetrics(workItems: WorkItem[], resourceLimit: number): ProjectMetrics {
  const schedule = isScheduled(workItems) ? (workItems as ScheduledWorkItem[]) : calculateSchedule(workItems).items;
  const usage = calculateResourceUsage(schedule, resourceLimit);
  const projectDuration = schedule.length ? Math.max(...schedule.map((item) => item.earlyFinish)) : 0;
  const maxWorkers = usage.length ? Math.max(...usage.map((point) => point.workers)) : 0;
  const averageWorkers = usage.length ? usage.reduce((sum, point) => sum + point.workers, 0) / usage.length : 0;
  const overloadDays = usage.filter((point) => point.overloaded).length;
  const idleDays = usage.filter((point) => point.idle).length;
  const efficiency = resourceLimit > 0 && projectDuration > 0 ? Math.round((averageWorkers / resourceLimit) * 100) : 0;

  return {
    projectDuration,
    maxWorkers,
    averageWorkers,
    criticalCount: schedule.filter((item) => item.isCritical).length,
    floatCount: schedule.filter((item) => item.totalFloat > 0).length,
    overloadDays,
    idleDays,
    efficiency,
  };
}

export function validateNetwork(workItems: WorkItem[]): NetworkValidationError[] {
  const errors: NetworkValidationError[] = [];
  const ids = new Set(workItems.map((item) => item.id));

  for (const item of workItems) {
    if (item.workers < 1) {
      errors.push({ type: 'invalid-workers', workId: item.id, message: `В работе ${item.code} сотрудников должно быть не меньше 1.` });
    }
    if (item.duration < 1) {
      errors.push({ type: 'invalid-duration', workId: item.id, message: `В работе ${item.code} длительность должна быть не меньше 1.` });
    }
    for (const dependencyId of item.dependencies) {
      if (!ids.has(dependencyId)) {
        errors.push({
          type: 'missing-dependency',
          workId: item.id,
          message: `Работа ${item.code} ссылается на несуществующую зависимость ${dependencyId}.`,
        });
      }
    }
  }

  if (errors.length === 0 && topologicalSort(workItems).length !== workItems.length) {
    errors.push({ type: 'cycle', field: 'project', message: 'В графике обнаружен цикл. Проверьте события начала и окончания.' });
  }

  return errors;
}

function sanitizePositiveNumber(value: number): number {
  return Number.isFinite(value) && value >= 1 ? value : 1;
}

function topologicalSort(workItems: WorkItem[]): string[] {
  const ids = new Set(workItems.map((item) => item.id));
  const incoming = new Map(workItems.map((item) => [item.id, item.dependencies.filter((id) => ids.has(id)).length]));
  const outgoing = buildSuccessors(workItems);
  const queue = workItems.filter((item) => incoming.get(item.id) === 0).map((item) => item.id);
  const order: string[] = [];

  while (queue.length > 0) {
    const id = queue.shift()!;
    order.push(id);

    for (const successorId of outgoing.get(id) ?? []) {
      incoming.set(successorId, (incoming.get(successorId) ?? 0) - 1);
      if (incoming.get(successorId) === 0) {
        queue.push(successorId);
      }
    }
  }

  return order;
}

function buildSuccessors(workItems: WorkItem[]): Map<string, string[]> {
  const successors = new Map(workItems.map((item) => [item.id, [] as string[]]));
  for (const item of workItems) {
    for (const dependencyId of item.dependencies) {
      successors.get(dependencyId)?.push(item.id);
    }
  }
  return successors;
}

function isScheduled(workItems: WorkItem[]): boolean {
  return workItems.every((item) => typeof item.earlyStart === 'number' && typeof item.earlyFinish === 'number');
}
